Add note to avoid requiring config/initializers [ci skip]

composerinteralia opened the pull request:
Closes #40021

`config/intializers` are loaded with `load` rather than `require` (and
always have been, since they were introduced way back in c6d6082562). So
explicitly requiring them can be problematic if they are not idempotent.

Adding this note will give us something to point to if somebody runs
into this again.

p8 commented:
Having 3 notes and 1 tip that are larger than the actual text is probably a sign that `initializers` cause a lot of confusion 😄.
Maybe the second one about subfolders can be plain text instead, with an example folder structure.

The last tip also applies to #40021 I think. If both files were merged into a single file there would be no need to require one from the other.

Maybe we should also make it more explicit what does not belong in `config/initializers`.
Should monkey patches like #40021 belong in `lib/` instead?

I also think it's never a good idea to require any file in `config/` yourself, not just from `config/initializers/`. Rails should be responsible for this.

composerinteralia commented:
Thanks for the review!

> Having 3 notes and 1 tip that are larger than the actual text is probably a sign that initializers cause a lot of confusion 😄.
Maybe the second one about subfolders can be plain text instead, with an example folder structure.

I think you are right that the structure of this section is a bit off. I will play around with the structure a bit and see if I can make things more clear.

> Maybe we should also make it more explicit what does not belong in config/initializers. Should monkey patches like #40021 belong in lib/ instead?

I didn't look to closely at the actual content of the files. Yeah, I agree that probably doesn't belong in config/initializers. Capturing what shouldn't go into the directory would be difficult, since there isn't really any limit to what shouldn't go there (models, migrations, JS files, PDFs of your tax returns). Capturing what **should** got there makes sense, and I think we already do that to some extent with:

> You can use initializers to hold configuration settings that should be made after all of the frameworks and gems are loaded, such as options to configure settings for these parts.

Do you think there is something we could add to make that more clear?

> I also think it's never a good idea to require any file in config/ yourself, not just from config/initializers/, even if Rails used require for loading initializers. Rails should be responsible for this.

That may be true, but seems a little out of scope for this particular section of the documentation. It's possible there are other areas of the documentation that could be made more clear.

p8 commented:
> Do you think there is something we could add to make that more clear?

Maybe that is good enough. We also have a description of what all directories are for in the getting_started guide.
We could maybe make it a bit more prescriptive:

>_Use_ initializers to hold configuration settings that should be made after all of the frameworks and gems are loaded, such as options to configure settings for these parts.

> That may be true, but seems a little out of scope for this particular section of the documentation. It's possible there are other areas of the documentation that could be made more clear.

Yes, I wasn't sure if we should make it more generic.

composerinteralia committed:
Add note to avoid requiring config/initializers

[ci skip]

Closes #40021

I began this commit by adding an additional note about explicitly
requiring initializers. `config/intializers` are loaded with `load`
rather than `require` (and always have been, since they were introduced
way back in c6d6082562). So explicitly requiring them can be problematic
if they are not idempotent.

But @p8 pointed out that it is a bit awkward to have 3 notes and one tip
larger than the actual text in this section.

This commit now also reworks 2 of the notes and the tip into additional
paragraphs in the main text. The benefits of the new version I see:

- We now explicitly say that the initializers are loaded by sort order
- We now start with the ideal way to deal with dependent initializers,
  then follow with a couple warnings. We used to lead with what not to
  do, which I think was a bit confusing.
- The language around subdirectories is much simpler (mentioning that
  subdirectories can be used for organizing seemed unnecessary to me -
  that is true of subdirectories in general, and doesn't have any
  specific connection to initializers)

This commit does not change the opening paragraph or the note about when
to use `after_initialize` instead, although it does reformat them.

composerinteralia commented:
@p8 I reworked this section a bit. Let me know what you think.

rafaelfranca reviewed (approved):
:shipit:

p8 commented:
Great improvement @composerinteralia !
